lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,351 @@
+  ) {
+    // Element not displayed
+    return false;
+  }
+
+  // only for block and inline-block, since clientWidth/Height are always 0 for inline elements
+  if (display === 'block' || display === 'inline-block') {
+    // if height/width is 0 and no overflow in that direction then
+    // there's no content that the user can see and tap on
+    if ((node.clientWidth === 0 && overflowX === 'hidden') ||
+        (node.clientHeight === 0 && overflowY === 'hidden')) {
+      return false;
+    }
+  }
+
+  const parent = node.parentElement;
+  if (parent && parent.tagName !== 'BODY' && !nodeIsVisible(parent)) {
+    // if a parent is invisible then the current node is also invisible
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * @param {LH.Artifacts.Rect[]} clientRects
+ */
+/* istanbul ignore next */
+function allClientRectsEmpty(clientRects) {
+  return clientRects.every(cr => cr.width === 0 && cr.height === 0);
+}
+
+/**
+ * @param {Element} node
+ */
+/* istanbul ignore next */
+function getVisibleClientRects(node) {
+  if (!nodeIsVisible(node)) {
+    return [];
+  }
+
+  const {
+    overflowX,
+    overflowY,
+  } = getComputedStyle(node);
+  let clientRects = getClientRects(node, true);
+
+  if (allClientRectsEmpty(clientRects)) {
+    if ((overflowX === 'hidden' && overflowY === 'hidden') || node.children.length === 0) {
+      // own size is 0x0 and there's no visible child content
+      return [];
+    }
+  }
+
+  // Treating overflowing content in scroll containers as invisible could mean that
+  // most of a given page is deemed invisible. But:
+  // - tap targets audit doesn't consider different containers/layers
+  // - having most content in an explicit scroll container is rare
+  // - treating them as hidden only generates false passes, which is better than false failures
+  clientRects = filterClientRectsWithinAncestorsVisibleScrollArea(node, clientRects);
+
+  return clientRects;
+}
+
+/**
+ *
+ * @param {Element} node
+ * @param {LH.Artifacts.Rect[]} clientRects
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function filterClientRectsWithinAncestorsVisibleScrollArea(node, clientRects) {
+  const parent = node.parentElement;
+  if (!parent) {
+    return clientRects;
+  }
+  if (getComputedStyle(parent).overflowY !== 'visible') {
+    const parentBCR = parent.getBoundingClientRect();
+    clientRects = clientRects.filter(cr => rectContains(parentBCR, cr));
+  }
+  if (parent.parentElement && parent.parentElement.tagName !== 'BODY') {
+    return filterClientRectsWithinAncestorsVisibleScrollArea(
+      parent,
+      clientRects
+    );
+  }
+  return clientRects;
+}
+
+/**
+ * @param {Element} node
+ * @param {boolean} includeChildren
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function getClientRects(node, includeChildren = true) {
+  /** @type {LH.Artifacts.Rect[]} */
+  let clientRects = Array.from(
```

Comment:
nit: can be `const` if the `clientRects.concat` below switches to `clientRects.push(...getClientRects(child))`